File: task-03/C++/Sub-Task4.cpp
#include <iostream>
#include <fstream>

void generateDiamond(int n, std::ofstream& outputFile) {
    // Top half including the middle line
    for (int i = 0; i < n; ++i) {
        // Print leading spaces
        for (int j = 0; j < n - i - 1; ++j) {
            outputFile << " ";
        }
        // Print asterisks
        for (int j = 0; j < 2 * i + 1; ++j) {
            outputFile << "*";
        }
        outputFile << std::endl;
    }

    // Bottom half excluding the middle line
    for (int i = n - 2; i >= 0; --i) {
        // Print leading spaces
        for (int j = 0; j < n - i - 1; ++j) {
            outputFile << " ";
        }
        // Print asterisks
        for (int j = 0; j < 2 * i + 1; ++j) {
            outputFile << "*";
        }
        outputFile << std::endl;
    }
}

int main() {
    int n;
    std::ifstream inputFile("input.txt");  // Open input.txt for reading
    std::ofstream outputFile("output.txt"); // Open output.txt for writing

    // Check if input file opened successfully
    if (!inputFile.is_open()) {
        std::cerr << "Error opening input.txt" << std::endl;
        return 1;
    }

    // Check if output file opened successfully
    if (!outputFile.is_open()) {
        std::cerr << "Error opening output.txt" << std::endl;
        return 1;
    }

    // Read the number from input.txt
    if (!(inputFile >> n)) {
        std::cerr << "Error reading the number from input.txt" << std::endl;
        return 1;
    }
    inputFile.close();

    // Validate the number
    if (n <= 0) {
        std::cerr << "The number must be a positive integer." << std::endl;
        return 1;
    }

    // Generate the diamond pattern and write it to output.txt
    generateDiamond(n, outputFile);
    outputFile.close();

    std::cout << "Diamond pattern generated successfully in output.txt." << std::endl;

    return 0;
}

File: task-03/Haskell/Sub-Task4.hs
import System.IO

main :: IO ()
main = do
    -- Read the number 'n' from input.txt
    inputFile <- openFile "input.txt" ReadMode
    input <- hGetLine inputFile
    let n = read input :: Int
    hClose inputFile

    -- Generate the diamond pattern
    let diamondPattern = generateDiamond n

    -- Write the diamond pattern to output.txt
    outputFile <- openFile "output.txt" WriteMode
    hPutStr outputFile diamondPattern
    hClose outputFile

-- Function to generate the diamond pattern
generateDiamond :: Int -> String
generateDiamond n = unlines $ upperHalf ++ lowerHalf
  where
    upperHalf = [generateLine i n | i <- [1..n]]
    lowerHalf = [generateLine i n | i <- [n-1,n-2..1]]

-- Function to generate a single line of the diamond
generateLine :: Int -> Int -> String
generateLine i n = replicate (n - i) ' ' ++ replicate (2 * i - 1) '*'
